quic/s2n-quic/src/provider/event/console_perf.rs

```diff
@@ -0,0 +1,296 @@
+        self.counters
+            .max_rtt
+            .fetch_max(event.latest_rtt.as_nanos() as _, Ordering::Relaxed);
+        self.counters
+            .max_smoothed_rtt
+            .fetch_max(event.smoothed_rtt.as_nanos() as _, Ordering::Relaxed);
+        self.counters
+            .pto_count
+            .fetch_max(event.pto_count as _, Ordering::Relaxed);
+    }
+
+    #[inline]
+    fn on_packet_lost(
+        &mut self,
+        _context: &mut Self::ConnectionContext,
+        _meta: &event::ConnectionMeta,
+        _event: &event::events::PacketLost,
+    ) {
+        self.counters.lost_packets.fetch_add(1, Ordering::Relaxed);
+    }
+
+    #[inline]
+    fn on_platform_event_loop_wakeup(
+        &mut self,
+        _meta: &event::events::EndpointMeta,
+        _event: &event::events::PlatformEventLoopWakeup,
+    ) {
+        self.counters
+            .event_loop_wakeup
+            .fetch_add(1, Ordering::Relaxed);
+    }
+
+    #[inline]
+    fn on_platform_event_loop_sleep(
+        &mut self,
+        _meta: &event::events::EndpointMeta,
+        event: &event::events::PlatformEventLoopSleep,
+    ) {
+        self.counters.timeout.store(
+            event.timeout.unwrap_or_default().as_nanos() as _,
+            Ordering::Relaxed,
+        );
+    }
+
+    #[inline]
+    fn on_pacing_rate_updated(
+        &mut self,
+        _context: &mut Self::ConnectionContext,
+        _meta: &event::events::ConnectionMeta,
+        event: &event::events::PacingRateUpdated,
+    ) {
+        self.counters
+            .max_pacing_rate
+            .fetch_max(event.bytes_per_second, Ordering::Relaxed);
+    }
+
+    #[inline]
+    fn on_delivery_rate_sampled(
+        &mut self,
+        _context: &mut Self::ConnectionContext,
+        _meta: &event::events::ConnectionMeta,
+        event: &event::events::DeliveryRateSampled,
+    ) {
+        self.counters.max_delivery_rate.fetch_max(
+            event.rate_sample.delivery_rate_bytes_per_second,
+            Ordering::Relaxed,
+        );
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct Counters {
+    send_progress: AtomicU64,
+    receive_progress: AtomicU64,
+    max_cwnd: AtomicU64,
+    max_bytes_in_flight: AtomicU64,
+    max_rtt: AtomicU64,
+    max_smoothed_rtt: AtomicU64,
+    lost_packets: AtomicU64,
+    event_loop_wakeup: AtomicU64,
+    timeout: AtomicU64,
+    pto_count: AtomicU64,
+    max_pacing_rate: AtomicU64,
+    max_delivery_rate: AtomicU64,
+}
+
+impl Counters {
+    pub fn print_header(&self) {
+        println!(
+            "Tx Rate\t\
+            Rx Rate\t\
+            Max Cwnd\t\
+            Max Inflight\t\
+            Lost Packets\t\
+            Wakeups\t\
+            Duration\t\
+            Max RTT\t\
+            Max SRTT\t\
+            PTO Count\t\
+            Max Pacing Rate\t\
+            Max Delivery Rate"
+        );
+    }
+
+    pub fn print(&self, duration: Duration) {
```

Comment:
Maybe we should call this `print_tsv` to make it clear what the format is. Or have `print(&self, duration: Duration, format: Format)` where `enum Format { TSV, JSON, }` in case we want to add new formats in the future.

Reply:
I could put the format in the struct

Reply:
That works